src/modules/bulletproofs/bulletproofs_pp_norm_product_impl.h

```diff
@@ -0,0 +1,589 @@
+        memcpy(&pk_buf[1], &data->proof[65*idx + 33], 32);
+        if (!secp256k1_eckey_pubkey_parse(pt, pk_buf, sizeof(pk_buf))) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+typedef struct ec_mult_verify_cb_data2 {
+    const secp256k1_scalar *s_g;
+    const secp256k1_scalar *s_h;
+    const secp256k1_ge *g_vec;
+    size_t g_vec_len;
+} ec_mult_verify_cb_data2;
+
+static int ec_mult_verify_cb2(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *cbdata) {
+    ec_mult_verify_cb_data2 *data = (ec_mult_verify_cb_data2*) cbdata;
+    if (idx < data->g_vec_len) {
+        *sc = data->s_g[idx];
+    } else {
+        *sc = data->s_h[idx - data->g_vec_len];
+    }
+    *pt = data->g_vec[idx];
+    return 1;
+}
+
+/* Verify the proof */
+int secp256k1_bulletproofs_pp_rangeproof_norm_product_verify(
+    const secp256k1_context* ctx,
+    secp256k1_scratch_space* scratch,
+    unsigned char* proof,
+    size_t proof_len,
+    unsigned char *transcript_hash32,
+    const secp256k1_scalar* r_ch,
+    secp256k1_bulletproofs_generators* g_vec,
+    size_t g_len,
+    secp256k1_scalar* c_vec,
+    size_t c_vec_len,
+    const secp256k1_ge* commit
+) {
+    secp256k1_scalar q, r, v, n, l, r_inv, h_c;
+    secp256k1_scalar *es, *s_g, *s_h, *r_inv_pows;
+    secp256k1_gej res1, res2;
+    size_t i = 0, scratch_checkpoint;
+    int overflow;
+    secp256k1_sha256 sha256;
+    unsigned char ser_commit[33];
+    size_t log_n = secp256k1_bulletproofs_pp_log2(g_len), log_m = secp256k1_bulletproofs_pp_log2(c_vec_len);
+    size_t n_rounds = log_n > log_m ? log_n : log_m;
+    size_t h_len = c_vec_len;
+    secp256k1_ge comm = *commit;
+
+    if (g_vec->n != (h_len + g_len) || (proof_len != 65 * n_rounds + 64)) {
+        return 0;
+    }
+
+    if (!secp256k1_check_power_of_two(g_len) ||  !secp256k1_check_power_of_two(h_len)) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_b32(&n, &proof[n_rounds*65], &overflow); /* n */
+    if (overflow) return 0;
+    secp256k1_scalar_set_b32(&l, &proof[n_rounds*65 + 32], &overflow); /* l */
+    if (overflow) return 0;
+    if (secp256k1_scalar_is_zero(r_ch)) return 0;
+
+    /* Compute powers of q_inv. Later used in g_factor computations*/
+    r = *r_ch;
+    secp256k1_scalar_inverse_var(&r_inv, &r);
+    scratch_checkpoint = secp256k1_scratch_checkpoint(&ctx->error_callback, scratch);
+
+    r_inv_pows = (secp256k1_scalar*)secp256k1_scratch_alloc(&ctx->error_callback, scratch, log_n * sizeof(secp256k1_scalar));
+    secp256k1_bulletproofs_powers_of_r(r_inv_pows, &r_inv, log_n);
+
+    secp256k1_sha256_initialize(&sha256);
+    secp256k1_sha256_write(&sha256, transcript_hash32, 32);
+    secp256k1_fe_normalize_var(&comm.x);
+    secp256k1_fe_normalize_var(&comm.y);
+    ser_commit[0] = 0x02 | secp256k1_fe_is_odd(&comm.y);
+    secp256k1_fe_get_b32(&ser_commit[1], &comm.x);
+    secp256k1_sha256_write(&sha256, ser_commit, 33);
+    secp256k1_sha256_finalize(&sha256, transcript_hash32);
```

Comment:
I was quickchecking the rust implementation and noticed that X and R can become the point at infinity. @Liam-Eagen and I found multiple conditions in which this can happen, e.g. when l and n don't have the same length and the longer one is the zero vector. I added a test case to this C code to demonstrate this [here](https://github.com/jonasnick/secp256k1-zkp/commit/2cbfe52571d9f200097158f341cd05465197cef6#diff-fd6906ded5f921ba870d059568d0d8a2ca9bc495d09e0738557cce6008603d3fR256).

As the added test also demonstrates, if X or R are the point at infinity then proving succeeds but verifying fails. That's because the prover algorithm serializes some invalid X-coordinate into the proof when X or R are the point at infinity. The verifier calls `eckey_parse` on the proof data which fails on invalid X-coordinates.

If we want to avoid that, I see two options:
1. make the prover fail when X or R are zero and add a note to the function documentation that either says that the function may fail when the inputs are not uniformly random or state the exact conditions when it happens
2. properly parse and serialize the point at infinity (the musig PR has some utilities for doing that)

I'd prefer option 2 because it doesn't add an error branch and doesn't complicate the doc. I've also updated the rust code to implement option 2.

Reply:
I also prefer option 2. We've had problems in past crypto (surjection proofs) where we discovered "legitimate" uses of intermediate infinity values but weren't able to use them ... and more annoyingly, being able to "successfully" create proofs for them that wouldn't verify.

Reply:
Sounds good to me!

Reply:
Fine for me. I've updated the tests that exercise this branch accordingly.